### ISSUE TITLE
TF Micro requires CONV_2D version '2' when applying quantization

### ISSUE THREAD

**Issue by dustedduke** [tensorflow/tensorflow#32468]
- Have I written custom code: Yes
- OS Platform and Distribution: Ubuntu 18.04
- TensorFlow version: 1.14, 2.0-rc0, nightly-preview 2.0.0.dev20190911
- Python version: 3.7.4

I am trying to run a simple Keras model with TensorFlow version for microcontrollers (modified micro_vision example with several convolutions, dense layers and batch normalization). While the model works correctly when using floats, the quantized version generated with `tf.lite.Optimize.DEFAULT` outputs wrong values and `'Didn't find op for builtin opcode 'CONV_2D' version '2' Invoke failed.'` error.

After changing `CONV_2D` version with `AddBuiltin(BuiltinOperator_CONV_2D, Register_CONV_2D(), 1, 2)` in `all_ops_resolver.cc` the error disappeared, but now the model outputs NaN instead of normal output and I'm not sure if it actually uses a different micro kernel.

After trying multiple conversion techniques (from .h5, model itself, SavedModel folder etc.) with multiple TF versions and BatchNormalization removal, the situation remains the same. As far as I understand, each way to quantize the model (post-training quantization, quantization aware training) requires this version of `CONV2D`, but it seems to work incorrectly or there is no such version in TF micro.

Is there any way to deploy such a model with quantization?

**Comment by rockyrhodes** [comment]
Thanks for this information.  I believe there are several problems brought to light here.  First, the appropriate version of the convolution op that is generated by this quantization flow is not yet available in the micro framework.  Float and uint8 versions of the op are currently supported, but the new tf.lite quantization flow generates int8 ops.  There is a change in process to add this op to the micro framework, but it is not currently supported.

Second, if you were able to get the error to disappear by making your change, but the model just silently fails, this is a problem as well.  A proper error should still be generated indicating a mismatch between the tensor type and the kernel versions available.  I'll look into this a bit more and see if I can understand what is happening.

**Comment by rockyrhodes** [comment]
The Eval(...) method in conv.cc is checking the input type of the op and generating an error message if the type is not one of {kTfLiteFloat32, kTfLiteUInt8}.  Do you see an error message like "Type %d not currently supported" when you run the code with your all_ops_resolver.cc changes?  The micro_interpreter code should also generate an error like "Node %s (number %d) failed to invoke with status %d" when the underlying invoke(...) method of an op returns an error.  The MicroInterpreter.Invoke(...) method should also return an error status.  Do you see this in your example code?

**Comment by dustedduke** [comment]
@rockyrhodes No, I see only CONV2D version error when using post-training quantization,

**Comment by njeffrie** [comment]
The error you're running into is definitely caused by us not having an int8 version (version 2) of CONV2D.  I'm working to add int8 per-channel quantization to DEPTHWISE_CONV and CONV2D, which should solve your problem.  This should land within the next week.  I would expect to see the type error that Rocky mentioned, so I'll have a quick look next week to make sure we are logging type errors correctly.

**Comment by csukuangfj** [comment]
@njeffrie 
I encountered the same problem with TF 1.14. And when I switched to TF 1.13.1 I have the following problem https://github.com/tensorflow/tensorflow/issues/32422

Could you please have a look at it?

**Comment by njeffrie** [comment]
Thanks for the reminder and apologies for the delay.  I'll have a look tomorrow.

**Comment by apparvi** [comment]
I see that @njeffrie pushed the int8 quantization support. I'm however still getting the same `'Didn't find op for builtin opcode'CONV_2D' version '2' Invoke failed.'` error with a simple convolutional network. 

I'm using the latest nightly build:
- Have I written custom code: Yes
- OS Platform and Distribution: Ubuntu 18.04
- TensorFlow version: 1.14, 2.0-rc0, nightly-preview 2.0.0-dev20191002
- Python version: 3.7.4

Any news @dustedduke? Were you able to deploy the quantized model?

**Comment by dustedduke** [comment]
@apparvi `CONV_2D` error doesn't appear in tf-nightly 2.1.0.dev20191007, but now I encounter different errors with recently added `QUANTIZE` layer and `DEPTHWISE_CONV_2D`.

When I try to generate a fully integer model, as explained [here](https://www.tensorflow.org/lite/performance/post_training_integer_quant), converted CNN starts asking for `QUANTIZE` layer (`Didn't find op for builtin opcode...`), and asks to change the input to float32 after adding the layer, although I manually set `converter.inference_input_type = tf.int8` or `converter.inference_input_type = tf.uint8`. The situation is different, when using hybrid model with float32 input: it successfully uses `QUANTIZE` layer, but then asks for `DEPTHWISE_CONV_2D version 3`. @njeffrie I'm wondering if converter tries to replace basic convolution with depthwise version for better performance, because `DEPTHWISE_CONV_2D` dependency is not mentioned in any of the example files and isn't used in the model.

**Comment by njeffrie** [comment]
It's weird that you're seeing `DEPTHWISE_CONV_2D` if your model does not use it.  We do not support hybrid quantization on micro, so I would expect to see that fail.

I notice that I forgot to increment the max op version in all_ops_resolver.  I will be working on streamlining how we do op versioning on micro soon.

@suharshs may be better able to help with the conversion step.

**Comment by dustedduke** [comment]
@suharshs As I understood, TF Lite conversion process replaces **first** `CONV_2D` with `DEPTHWISE_CONV_2D` with depth multiplier = number of output channels (16 in my case). Now, after running the model with added depthwise operation, I see the following:

```
tensorflow/lite/kernels/kernel_util.cc:54 affine_quantization->scale->size != filter->dims->data[affine_quantization->quantized_dimension] (16 != 1) 
```
Is it a conversion issue, or I can set it manually somewhere?

**Comment by stnk20** [comment]
I also encountered the same problem.
I found that this problem seems to be happening because tfmicro does not have implementation of  "requantization"　(introduced by commit a1d8d4a33529c1dff642177302dd19c2dfd06c1d).

**Comment by AdamMoses-GitHub** [comment]
Any further updates on this, I'm seeing:

> Didn't find op for builtin opcode 'CONCATENATION' version '1'
> Failed to get registration from op code  d

**Comment by metanav** [comment]
I am having the same issue. Please add missing ops for quantized model.

**Comment by AdamMoses-GitHub** [comment]
I was able to overcome this by cloning the latest version of Tensorflow and building just the examples from TFLite Micro directories. Once I had that I could map my CMakeLists to the same makefile arrangement that was used there, copied the same source tree, and it worked.

I think in the past I had been using TF 1.12, as that's what my model was builtin, but once I upgraded everything to the TF master copy it has been fine. The number of operations being added in just the last few iterations has been numerous.

**Comment by metanav** [comment]
@AdamMoses-GitHub  Are you talking about CONCATENATION version '1' only or they have added  CONV_2D version '2', too?

**Comment by ShawnHymel** [comment]
Has there been any updates on this? I'm running into the same error as @dustedduke . I ran the example found here: https://www.tensorflow.org/lite/performance/post_training_integer_quant to produce the *mnist_model_quant_io.tflite* model.

I ran that model through this converter code to create a .h file with the model as constant bytes:

```
# Function: Convert some hex value into an array for C programming
def hex_to_c_array(hex_data, var_name):
    
    c_str = ''
    
    # Create header guard
    c_str += '#ifndef ' + var_name.upper() + '_H\n'
    c_str += '#define ' + var_name.upper() + '_H\n\n'
    
    # Add array length at top of file
    c_str += '\nunsigned int ' + var_name + '_len = ' + str(len(hex_data)) + ';\n'
    
    # Declare C variable
    c_str += 'unsigned char ' + var_name + '[] = {'
    hex_array = []
    for i, val in enumerate(hex_data):

        # Construct string from hex
        hex_str = format(val, '#04x')
        
        # Add formatting so each line stays within 80 characters
        if (i + 1) < len(hex_data):
            hex_str += ','
        if (i + 1) % 12 == 0:
            hex_str += '\n '
        hex_array.append(hex_str)
        
    # Add closing brace
    c_str += '\n  ' + format(' '.join(hex_array)) + '\n};\n\n'
    
    # Close out header guard
    c_str += '#endif //' + var_name.upper() + '_H'
    
    return c_str

# Write TFLite model to a C source file
c_model_name = "mnist_model_quant_io"
with open(c_model_name + '.h', 'w') as file:
    file.write(hex_to_c_array(tflite_model_quant, c_model_name))
```

I loaded the file on Arduino (TensorFlow library v1.15.0-ALPHA-precompiled) and attempted to run inference. I got the following error:

```
/home/arduino/workspace/Libraries - Google-Tensorflow scraper/Arduino/libraries/tensorflow_lite_mirror/src/tensorflow/lite/kernels/kernel_util.cpp:54 affine_quantization->scale->size != filter->dims->data[affine_quantization->quantized_dimension] (4 != 1)
Node DEPTHWISE_CONV_2D (number 2) failed to invoke with status 1
```

Here is the model from the demo:

```
model = keras.Sequential([
  keras.layers.InputLayer(input_shape=(28, 28)),
  keras.layers.Reshape(target_shape=(28, 28, 1)),
  keras.layers.Conv2D(filters=4, kernel_size=(2, 2), activation=tf.nn.relu),
  keras.layers.MaxPooling2D(pool_size=(4, 4)),
  keras.layers.Flatten(),
  keras.layers.Dense(10, activation=tf.nn.softmax)
])
```

As you can see, there is no DepthwiseConv2D in the Keras model. It looks like the TFLite converter is replacing the first Conv2D layer with a DepthwiseConv2D layer and changing the depth_multiplier to the same number as the filters. This seems to break things, but only when we add quantization (it works without quantization).

Any help with this would be appreciated!

Python 3.7.6
TensorFlow 2.1.0
Keras 2.2.4-tf
Windows 10 x64
